tls/s2n_prf.c

```diff
@@ -459,43 +461,129 @@ S2N_RESULT s2n_prf_free(struct s2n_connection *conn)
     return S2N_RESULT_OK;
 }
 
-static int s2n_prf(struct s2n_connection *conn, struct s2n_blob *secret, struct s2n_blob *label, struct s2n_blob *seed_a,
-        struct s2n_blob *seed_b, struct s2n_blob *seed_c, struct s2n_blob *out)
+static bool s2n_libcrypto_supports_tls_prf()
 {
-    POSIX_ENSURE_REF(conn);
-    POSIX_ENSURE_REF(secret);
-    POSIX_ENSURE_REF(conn->prf_space);
-    POSIX_ENSURE_REF(conn->secure);
-
-    /* seed_a is always required, seed_b is optional, if seed_c is provided seed_b must also be provided */
-    S2N_ERROR_IF(seed_a == NULL, S2N_ERR_PRF_INVALID_SEED);
-    S2N_ERROR_IF(seed_b == NULL && seed_c != NULL, S2N_ERR_PRF_INVALID_SEED);
-
-    if (conn->actual_protocol_version == S2N_SSLv3) {
-        return s2n_sslv3_prf(conn, secret, seed_a, seed_b, seed_c, out);
-    }
+    return S2N_LIBCRYPTO_SUPPORTS_TLS_PRF;
+}
 
+static S2N_RESULT s2n_custom_prf(struct s2n_connection *conn, struct s2n_blob *secret, struct s2n_blob *label, struct s2n_blob *seed_a,
+        struct s2n_blob *seed_b, struct s2n_blob *seed_c, struct s2n_blob *out)
+{
     /* We zero the out blob because p_hash works by XOR'ing with the existing
      * buffer. This is a little convoluted but means we can avoid dynamic memory
      * allocation. When we call p_hash once (in the TLS1.2 case) it will produce
      * the right values. When we call it twice in the regular case, the two
      * outputs will be XORd just ass the TLS 1.0 and 1.1 RFCs require.
      */
-    POSIX_GUARD(s2n_blob_zero(out));
+    RESULT_GUARD_POSIX(s2n_blob_zero(out));
 
     if (conn->actual_protocol_version == S2N_TLS12) {
-        return s2n_p_hash(conn->prf_space, conn->secure->cipher_suite->prf_alg, secret, label, seed_a, seed_b,
-                seed_c, out);
+        RESULT_GUARD_POSIX(s2n_p_hash(conn->prf_space, conn->secure->cipher_suite->prf_alg, secret, label, seed_a, seed_b, seed_c, out));
+        return S2N_RESULT_OK;
     }
 
     struct s2n_blob half_secret = { 0 };
-    POSIX_GUARD(s2n_blob_init(&half_secret, secret->data, (secret->size + 1) / 2));
+    RESULT_GUARD_POSIX(s2n_blob_init(&half_secret, secret->data, (secret->size + 1) / 2));
 
-    POSIX_GUARD(s2n_p_hash(conn->prf_space, S2N_HMAC_MD5, &half_secret, label, seed_a, seed_b, seed_c, out));
+    RESULT_GUARD_POSIX(s2n_p_hash(conn->prf_space, S2N_HMAC_MD5, &half_secret, label, seed_a, seed_b, seed_c, out));
     half_secret.data += secret->size - half_secret.size;
-    POSIX_GUARD(s2n_p_hash(conn->prf_space, S2N_HMAC_SHA1, &half_secret, label, seed_a, seed_b, seed_c, out));
+    RESULT_GUARD_POSIX(s2n_p_hash(conn->prf_space, S2N_HMAC_SHA1, &half_secret, label, seed_a, seed_b, seed_c, out));
 
-    return 0;
+    return S2N_RESULT_OK;
+}
+
+#if S2N_LIBCRYPTO_SUPPORTS_TLS_PRF
+/* BoringSSL and AWSLC define the CRYPTO_tls1_prf API in a private header. This function is
+ * forward-declared to make it accessible.
+ */
```

Comment:
Do you have any concerns about visibility / this method not being stable? For example, this would be a very bad idea to do with any s2n-tls method not officially in the api.

Reply:
I added a comment to explain this. I also removed the feature probe. After testing the probe by rebuilding AWSLC with modified API definitions, the try_compile was not sufficient to catch the changes and disable the API. If the modification just hid the symbol, the try_compile worked, but if an argument was added or something the try_compile failed to catch it. Since BoringSSL could theoretically change this API definition, I gated it behind just AWSLC rather than a feature probe.

Reply:
And just to confirm, since we're not using a try_compiler, the method is available and exported in all versions of awslc? Even the oldest?

Reply:
Correct - `CRYPTO_tls1_prf` was made visible by https://github.com/aws/aws-lc/commit/37c6eb4284bea36be9fa41d35b582008f2023dcf in 2018. The first AWSLC release was [v0.1-alpha](https://github.com/aws/aws-lc/releases/tag/v0.1-alpha), released in 2020. The API is exported in this version here: https://github.com/aws/aws-lc/blob/v0.1-alpha/third_party/boringssl/crypto/fipsmodule/tls/internal.h#L27-L33